[SPARK-15652] [launcher] Added a new State (LOST) for the listeners of SparkLauncher

subrotosanyal opened the pull request:
## What changes were proposed in this pull request?

This situation can happen when the LauncherConnection gets an exception while reading through the socket and terminating silently without notifying making the client/listener think that the job is still in previous state.
The fix force sends a notification to client that the job finished with unknown status and let client handle it accordingly.
## How was this patch tested?

Added a unit test.

Commit:
SPARK-15652 Added a new State (FINISHED_UNKNOW) which will be used if the LauncherConnection gets an exception while reading through the socket. The failure of reading through the socket could be crash of the Spark Submit JVM. The fix force sends a notification to client that the job finished with unknown status.

AmplabJenkins commented:
Can one of the admins verify this patch?

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/LauncherConnection.java:
```diff
@@ -107,4 +106,17 @@ public void close() throws IOException {
   }
 
+  private void close(Exception exp) throws IOException {
+	if (!closed) {
```
Spark uses 2 spaces for indentation. Your code later is also indented wrongly.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/LauncherConnection.java:
```diff
@@ -107,4 +106,17 @@ public void close() throws IOException {
+	if (!closed) {
+	  synchronized (this) {
+	    if (!closed) {
+	      handle(new SetState(State.FINISHED_UNKNOWN));
```
Hmm, this looks out of place. I think a better place for this is the `ServerConnection` class (in `LauncherServer.java`). That's the class that understands this message. Then your patch is also much simpler since you don't need this second close method, it turns into a single line in `ServerConnection.close` (and maybe a log message).

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/SparkAppHandle.java:
```diff
@@ -43,6 +43,8 @@
     /** The application finished with a successful status. */
     FINISHED(true),
+    /** The Spark Submit JVM got killed with a unknown status. */
+    FINISHED_UNKNOWN(true),
```
This should be declared after existing constants, since declaring it here will change the `ordinal()` value of existing constants, which is a binary compatibility issue. A better name would be good also; maybe "LOST" or "DISCONNECTED"?

vanzin commented on launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java:
```diff
@@ -17,19 +17,42 @@
 
 package org.apache.spark.launcher;
 
+import static org.junit.Assert.assertEquals;
```
This might be your IDE, but please don't explode (or re-order) existing imports like this. It's unnecessary noise and actually against Spark's code conventions.

vanzin commented on launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java:
```diff
@@ -17,19 +17,42 @@
-import static org.junit.Assert.*;
+import org.mockito.Mockito;
+
+import com.google.common.collect.Lists;
```
This module doesn't depend on Guava, so this would probably cause a failure (at least when using maven). With Java 7 it's pretty easy to avoid these particular Guava classes, though.

vanzin commented on launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java:
```diff
@@ -152,6 +175,102 @@ public void testTimeout() throws Exception {
 
+  @Test
+  public void testSparkSubmitVmShutsDown() throws Exception {
+	LauncherServer launchServer = (LauncherServer)Mockito.mock(LauncherServer.class);
```
style nits:

- code is cleaner if you just static import `Mockito.*`
- space after casts: `(LauncherServer) mock(...)`

vanzin commented on launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java:
```diff
@@ -152,6 +175,102 @@ public void testTimeout() throws Exception {
   }
 
+  @Test
+  public void testSparkSubmitVmShutsDown() throws Exception {
```
This test looks very complicated. Why not:

- create a server and a client that connects to it
- initiate the connection (send "Hello")
- close the client socket
- verify the server socket fires the expected event

This would be very similar to `testCommunication` above until about L112. After that you just close the client socket and wait for the event on the handle.

vanzin commented:
ok to test

vanzin commented:
@subrotosanyal you also need to change the title to follow the usual convention. thx

SparkQA commented:
**[Test build #59960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59960/consoleFull)** for PR 13497 at commit [`3d9b78f`](https://github.com/apache/spark/commit/3d9b78f4656a2e2329ee014199e7cc4fe42ac375).

SparkQA commented:
**[Test build #59960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59960/consoleFull)** for PR 13497 at commit [`3d9b78f`](https://github.com/apache/spark/commit/3d9b78f4656a2e2329ee014199e7cc4fe42ac375).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59960/
Test PASSed.

SparkQA commented:
**[Test build #60034 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60034/consoleFull)** for PR 13497 at commit [`52118e6`](https://github.com/apache/spark/commit/52118e63cded342565c7c477c7ec7e011fc63633).

SparkQA commented:
**[Test build #60034 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60034/consoleFull)** for PR 13497 at commit [`52118e6`](https://github.com/apache/spark/commit/52118e63cded342565c7c477c7ec7e011fc63633).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60034/
Test FAILed.

Commit:
SPARK-15652 Code review comments:
1) Simplify the test-case
2) Keeping the LauncherConnection clean
3) Coding guidelines and check

SparkQA commented:
**[Test build #60040 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60040/consoleFull)** for PR 13497 at commit [`f157c73`](https://github.com/apache/spark/commit/f157c733ad75ebc5235b2a7cb1d979f0429be943).

subrotosanyal commented:
hi @vanzin 
Worked on the comments. Request code review.

SparkQA commented:
**[Test build #60046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60046/consoleFull)** for PR 13497 at commit [`da229b0`](https://github.com/apache/spark/commit/da229b0cce0d69afe7f30220ed31329f46d65356).

SparkQA commented:
**[Test build #60047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60047/consoleFull)** for PR 13497 at commit [`0472e25`](https://github.com/apache/spark/commit/0472e2549a24fffe31c766a27765bf6c1c6895df).

SparkQA commented:
**[Test build #60048 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60048/consoleFull)** for PR 13497 at commit [`c27e2c3`](https://github.com/apache/spark/commit/c27e2c378ddb74a580c12433cfc33bf2cc16f8e2).

SparkQA commented:
**[Test build #60049 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60049/consoleFull)** for PR 13497 at commit [`1ba8c48`](https://github.com/apache/spark/commit/1ba8c4894ed12db6eb26feeb0544a1fa9c5b9b6d).

SparkQA commented:
**[Test build #60040 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60040/consoleFull)** for PR 13497 at commit [`f157c73`](https://github.com/apache/spark/commit/f157c733ad75ebc5235b2a7cb1d979f0429be943).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60040/
Test PASSed.

SparkQA commented:
**[Test build #60046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60046/consoleFull)** for PR 13497 at commit [`da229b0`](https://github.com/apache/spark/commit/da229b0cce0d69afe7f30220ed31329f46d65356).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60046/
Test FAILed.

SparkQA commented:
**[Test build #60047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60047/consoleFull)** for PR 13497 at commit [`0472e25`](https://github.com/apache/spark/commit/0472e2549a24fffe31c766a27765bf6c1c6895df).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60047/
Test FAILed.

SparkQA commented:
**[Test build #60048 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60048/consoleFull)** for PR 13497 at commit [`c27e2c3`](https://github.com/apache/spark/commit/c27e2c378ddb74a580c12433cfc33bf2cc16f8e2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60048/
Test PASSed.

SparkQA commented:
**[Test build #60049 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60049/consoleFull)** for PR 13497 at commit [`1ba8c48`](https://github.com/apache/spark/commit/1ba8c4894ed12db6eb26feeb0544a1fa9c5b9b6d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60049/
Test PASSed.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/LauncherServer.java:
```diff
@@ -337,6 +337,10 @@ public void close() throws IOException {
       super.close();
       if (handle != null) {
+    	if (!handle.getState().isFinal()) {
+    	  LOG.log(Level.WARNING, "Lost connection to launch server.");
```
minor: this is running in the server, so you lost the connection to the Spark application.

vanzin commented on launcher/src/main/java/org/apache/spark/launcher/SparkAppHandle.java:
```diff
@@ -46,7 +46,9 @@
     /** The application was killed. */
-    KILLED(true);
+    KILLED(true),
+    /** The Spark Submit JVM got killed with a unknown status. */
```
s/got killed/exited

vanzin commented on launcher/src/test/java/org/apache/spark/launcher/LauncherServerSuite.java:
```diff
@@ -152,6 +157,69 @@ public void testTimeout() throws Exception {
+      };
+      sparkLauncherClient.start();
+      close(client);
+      sparkLauncherClient.join(10000L);
```
Do you need this separate thread? Seems to me like all you need is:

```
close(client);
assertTrue(semaphore.tryAcquire(30, TimeUnit.SECONDS));
assertEquals(SparkAppHandle.State.LOST, handle.getState());
```

That gets rid of a bunch of code in your tests and tests just the code you're interested in.

Commit:
SPARK-15652 Code review comments:
1) Cleanup test-case
2) better documentation of enum
3) updation in log message

SparkQA commented:
**[Test build #60058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60058/consoleFull)** for PR 13497 at commit [`3edb36a`](https://github.com/apache/spark/commit/3edb36a363ef994c2866d224921f1545e825f7d9).

SparkQA commented:
**[Test build #60059 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60059/consoleFull)** for PR 13497 at commit [`72ea2ca`](https://github.com/apache/spark/commit/72ea2ca05a55d0cab5b8cd694df6f95fb9a0c76a).

SparkQA commented:
**[Test build #60058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60058/consoleFull)** for PR 13497 at commit [`3edb36a`](https://github.com/apache/spark/commit/3edb36a363ef994c2866d224921f1545e825f7d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60058/
Test PASSed.

SparkQA commented:
**[Test build #60059 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60059/consoleFull)** for PR 13497 at commit [`72ea2ca`](https://github.com/apache/spark/commit/72ea2ca05a55d0cab5b8cd694df6f95fb9a0c76a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60059/
Test PASSed.

subrotosanyal commented:
hi @vanzin 
Thanks for your comments. Could you please have look to the changes?
The build is green.

vanzin commented:
LGTM. Merging to master / 2.0.
